tests/41end-to-end-keys/08-cross-signing.pl

```diff
@@ -0,0 +1,642 @@
+             "keys" => {
+                 "curve25519:".$user2_device => "curve25519+key",
+                 "ed25519:".$user2_device => "ed25519+key",
+             },
+             "signatures" => {
+                 $user2_id => {
+                     "ed25519:".$user2_device => "self+signature",
+                 },
+             },
+      } )->then( sub {
+         matrix_create_room( $user1 );
+      })->then( sub {
+         ( $room_id ) = @_;
+
+         matrix_invite_user_to_room( $user1, $user2, $room_id )
+      })->then( sub {
+         matrix_sync( $user1 );
+      })->then( sub {
+         matrix_join_room( $user2, $room_id );
+      })->then( sub {
+         sync_until_user_in_device_list( $user1, $user2 );
+      })->then( sub {
+         matrix_set_cross_signing_key( $user2, {
+             "auth" => {
+                 "type"     => "m.login.password",
+                  "user"     => $user2_id,
+                  "password" => $user2->password,
+             },
+             "master_key" => {
+                 # private key: 2lonYOM6xYKdEsO+6KrC766xBcHnYnim1x/4LFGF8B0
+                 "user_id" => $user2_id,
+                 "usage" => ["master"],
+                 "keys" => {
+                     "ed25519:nqOvzeuGWT/sRx3h7+MHoInYj3Uk2LD/unI9kDYcHwk"
+                         => "nqOvzeuGWT/sRx3h7+MHoInYj3Uk2LD/unI9kDYcHwk",
+                 },
+             },
+             "self_signing_key" => $self_signing_key,
+         });
+      })->then( sub {
+         sync_until_user_in_device_list( $user1, $user2 );
+      })->then( sub {
+         matrix_get_keys( $user1, $user2_id );
+      })->then( sub {
+         my ( $content ) = @_;
+
+         log_if_fail "key query content", $content;
+
+         assert_json_keys( $content, "master_keys" );
+         assert_json_keys( $content->{master_keys}, $user2_id );
+         assert_deeply_eq( $content->{master_keys}->{$user2_id}, {
+               "user_id" => $user2_id,
+               "usage" => ["master"],
+               "keys" => {
+                  "ed25519:nqOvzeuGWT/sRx3h7+MHoInYj3Uk2LD/unI9kDYcHwk"
+                      => "nqOvzeuGWT/sRx3h7+MHoInYj3Uk2LD/unI9kDYcHwk",
+               },
+            },
+         );
+         assert_json_keys( $content, "self_signing_keys" );
+         assert_json_keys( $content->{self_signing_keys}, $user2_id );
+         assert_deeply_eq( $content->{self_signing_keys}->{$user2_id}, $self_signing_key);
+
+         Future->done(1);
+      });
+   };
+
+=head2 matrix_set_cross_signing_key
+
+   matrix_set_cross_signing_key( $user, $keys )
+
+Set cross-signing keys
+
+=cut
+
+sub matrix_set_cross_signing_key {
+   my ( $user, $keys ) = @_;
+
+   do_request_json_for( $user,
+      method  => "POST",
+      uri     => "/unstable/keys/device_signing/upload",
+      content => $keys,
+   );
+}
+
+=head2 matrix_get_keys
+
+   matrix_get_keys( $user, $keys )
+
+Get a user's keys
+
+=cut
+
+sub matrix_get_keys {
```

Comment:
this isn't device-keys specific, so probably belongs in `tests/41end-to-end-keys/01-upload-key.pl`